cpp/src/arrow/compute/kernels/scalar_compare_test.cc

```diff
@@ -1865,5 +1861,636 @@ TEST(TestMaxElementWiseMinElementWise, CommonTemporal) {
+                    R"(["よしもと","の","ち"])"),
+      ArrayFromJSON(TypeTraits<StringType>::type_singleton(), R"(["は","へ","あ"])"),
+      ArrayFromJSON(TypeTraits<StringType>::type_singleton(), R"(["な","を","ち"])"),
+      ArrayFromJSON(TypeTraits<BooleanType>::type_singleton(), "[false, false, true]"));
+  ValidateBetween<StringType>(
+      InclusiveOption,
+      ArrayFromJSON(TypeTraits<StringType>::type_singleton(), R"(["A","ア","王"])"),
+      ArrayFromJSON(TypeTraits<StringType>::type_singleton(), R"(["た","あ","歩"])"),
+      ArrayFromJSON(TypeTraits<StringType>::type_singleton(), R"(["李","田",null])"),
+      ArrayFromJSON(TypeTraits<BooleanType>::type_singleton(), "[false, true, null]"));
+  ValidateBetween<StringType>(
+      InclusiveOption,
+      ArrayFromJSON(TypeTraits<StringType>::type_singleton(), R"(["Б",null,"Я"])"),
+      ArrayFromJSON(TypeTraits<StringType>::type_singleton(), R"(["А","Ж","Щ"])"),
+      ArrayFromJSON(TypeTraits<StringType>::type_singleton(), R"(["Д","Л","Ф"])"),
+      ArrayFromJSON(TypeTraits<BooleanType>::type_singleton(), "[true, null, false]"));
+}
+
+TEST(TestTimestampsBetweenKernel, TimestampsArrayArrayArrayTest) {
+  const std::string arr_json = R"(["1970-01-01","2000-02-02","1900-02-28"])";
+  const std::string lhs_json = R"(["1970-01-01","2000-02-01","1900-02-28"])";
+  const std::string rhs_json = R"(["1970-01-02","2000-02-02","1900-02-28"])";
+  // Between Options
+  BetweenOptions both(BetweenOptions::Inclusive::BOTH);
+  BetweenOptions left(BetweenOptions::Inclusive::LEFT);
+  BetweenOptions right(BetweenOptions::Inclusive::RIGHT);
+  BetweenOptions neither(BetweenOptions::Inclusive::NEITHER);
+  // Same units should be fine
+  auto arr = ArrayFromJSON(timestamp(TimeUnit::SECOND), arr_json);
+  auto lhs = ArrayFromJSON(timestamp(TimeUnit::SECOND), lhs_json);
+  auto rhs = ArrayFromJSON(timestamp(TimeUnit::SECOND), rhs_json);
+  // BOTH
+  ValidateBetween<TimestampType>(both, arr, lhs, rhs);
+  // LEFT
+  ValidateBetween<TimestampType>(left, arr, lhs, rhs);
+  // RIGHT
+  ValidateBetween<TimestampType>(right, arr, lhs, rhs);
+  // NEITHER
+  ValidateBetween<TimestampType>(neither, arr, lhs, rhs);
+/*  // So are same time zones
+  arr = ArrayFromJSON(timestamp(TimeUnit::SECOND, "America/Chicago"), arr_json);
+  lhs = ArrayFromJSON(timestamp(TimeUnit::SECOND, "America/Chicago"), lhs_json);
+  rhs = ArrayFromJSON(timestamp(TimeUnit::SECOND, "America/Chicago"), rhs_json);
+  // BOTH
+  ValidateBetween<TimestampType>(both, arr, lhs, rhs);
+  // LEFT
+  ValidateBetween<TimestampType>(left, arr, lhs, rhs);
+  // RIGHT
+  ValidateBetween<TimestampType>(right, arr, lhs, rhs);
+  // NEITHER
+  ValidateBetween<TimestampType>(neither, arr, lhs, rhs);
+  // Different units should be fine
+  auto arr = ArrayFromJSON(timestamp(TimeUnit::SECOND), arr_json);
+  auto lhs = ArrayFromJSON(timestamp(TimeUnit::MILLI), lhs_json);
+  auto rhs = ArrayFromJSON(timestamp(TimeUnit::SECOND), rhs_json);
+  // BOTH
```

Comment:
The error here is because you are redeclaring `arr, lhs, rhs` which were already declared in the first case of this test. Simply remove the `auto` keyword from the above 3 statements and it should work correctly. I tested it locally. 😁

Reply:
cc @bkmgit

Reply:
Updating it

Reply:
Compare kernels for timestamps have a conversion of the form  `InputType in_type(match::TimestampTypeUnit(ty));` this is probably also helpful for Between.

Reply:
Some fixes are still needed to pass tests,  (timezones are fine, but timestamps with different units are not), otherwise most of the functionality is there now.